src/Aspire.Hosting/AspireStore.cs

```diff
@@ -0,0 +1,234 @@
+    public static AspireStore Create(IDistributedApplicationBuilder builder)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        var assemblyMetadata = builder.AppHostAssembly?.GetCustomAttributes<AssemblyMetadataAttribute>();
+        var objDir = GetMetadataValue(assemblyMetadata, "AppHostProjectBaseIntermediateOutputPath");
+
+        var fallbackDir = builder.Configuration[AspireStorePathKeyName];
+
+        var root = fallbackDir ?? objDir;
+
+        if (string.IsNullOrEmpty(root))
+        {
+            throw new InvalidOperationException($"Could not determine an appropriate location for storing user secrets. Set the {AspireStorePathKeyName} setting to a folder where the App Host content should be stored.");
+        }
+
+        var directoryPath = Path.Combine(root, ".aspire");
+
+        // The /obj directory doesn't need to be prefixed with the app host name.
+        if (root != objDir)
+        {
+            directoryPath = Path.Combine(directoryPath, GetAppHostSpecificPrefix(builder));
+        }
+
+        return new AspireStore(directoryPath);
+    }
+
+    /// <summary>
+    /// Gets the metadata value for the specified key from the assembly metadata.
+    /// </summary>
+    /// <param name="assemblyMetadata">The assembly metadata.</param>
+    /// <param name="key">The key to look for.</param>
+    /// <returns>The metadata value if found; otherwise, null.</returns>
+    private static string? GetMetadataValue(IEnumerable<AssemblyMetadataAttribute>? assemblyMetadata, string key) =>
+        assemblyMetadata?.FirstOrDefault(a => string.Equals(a.Key, key, StringComparison.OrdinalIgnoreCase))?.Value;
+
+    /// <summary>
+    /// Gets the application host specific prefix based on the builder's environment.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <returns>The application host specific prefix.</returns>
+    private static string GetAppHostSpecificPrefix(IDistributedApplicationBuilder builder)
+    {
+        var appName = Sanitize(builder.Environment.ApplicationName).ToLowerInvariant();
+        var appNameHash = builder.Configuration["AppHost:Sha256"]![..10].ToLowerInvariant();
+        return $"{appName}.{appNameHash}";
+    }
+
+    /// <summary>
+    /// Gets a deterministic file path that is a copy of the <paramref name="sourceFilename"/>.
+    /// The resulting file name will depend on the content of the file.
+    /// </summary>
+    /// <param name="filename">A file name to base the result on.</param>
+    /// <param name="sourceFilename">An existing file.</param>
+    /// <returns>A deterministic file path with the same content as <paramref name="sourceFilename"/>.</returns>
+    /// <exception cref="FileNotFoundException">Thrown when the source file does not exist.</exception>
+    public string GetFileNameWithContent(string filename, string sourceFilename)
```

Comment:
Should we call `sourceFilename` something like `sourceFilenameTemplate`? Since we will be appending the hash to it?